### PR TITLE
i18next

### PR DESCRIPTION
**Pull request by delphiactual**
#1767 

 - Grunt Tasks
   - [x] Automatically download updates for other languages.
   - [ ] Automatically update english terms.
 - angular-translate -> i18next
   - [x] All `translate=` -> `ng-18next=`
   - [x] All `$translate.instant` -> `i18next.t`
   - [x] Ensure all `contexts` work
   - [x] Ensure all `plurals` work
 - Current Snags
  - [x] ~~DIM no longer downloads selected language database... (assistance appreciated)~~ #1904
  - [x] MINOR: During initial load, all translations are in English(detected browser language)

i18next will provide for a much easier translation experience for the translators; the logic is no longer contained within the translation strings themselves but within i18next itself.

i18next/ng-i18next are also compatible with angular1 & angular whereas angular-translate is only available for angular1. Another translate service ng-translate is compatible with angular but does not have plural support as of yet...

key: "some key"
key_plural: "some plural key"  // pass in the reserved variable `count` to determine whether plural is used
key_male: "some male key"
key_female: "some female key" // pass in the reserved variable `context` that resolves to the correct context in this case `gender`

| | Added Module | size | Removed Module | size |
|-|-|-|-|-|
| |i18next | 79.55kb | angular-translate| 125.9kb |
| |angular-sanitize | 27.14kb | message-format |28.71kb |
| |ng-18next | 13.56kb |  angular-translate-interpolation-messageformat| 6.39kb|
|Total|| 120.25kb | |161kb |

Savings: 40.75kb

//cc @bhollis @SunburnedGoose @kyleshay

### CODE REVIEW

**Comment by delphiactual** [comment]
@bhollis for reviewing use https://github.com/DestinyItemManager/DIM/pull/1769/files?w=1

**Comment by bhollis** [comment]
Are there any bugs/tasks remaining that you know of? Last I checked there was still some problem with bootstrapping the right language?

**Comment by delphiactual** [comment]
Bootstrapping was fixed with #1904 so as soon as I get these merge conflicts solved this will be good to go.

File: src/scripts/dimApp.module.js [hunk 1]
```diff
@@ -35,17 +39,20 @@ import run from './dimApp.run';
 import state from './state';
 import loadingTracker from './services/dimLoadingTracker.factory';
 
+// required to make ng-i18next work
+window.i18next = i18next;
```
Comment (bhollis) [inline src/scripts/dimApp.module.js]: TODO: I actually know how to do this slightly more elegantly with webpack's `alias` now. No big deal though.

File: src/scripts/infuse/infuse.html [hunk 2]
```diff
@@ -18,34 +18,34 @@ <h1 translate="Infusion"></h1>
       </div>
 
       <div class="infuseControls">
-        <label><input class="controls" type="checkbox" id="infuse-all" ng-model="vm.getAllItems" ng-change="vm.getItems()"><span translate="Infusion.ShowItems"></span></label>
-        <label><input class="controls" type="checkbox" id="infuse-locked" ng-model="vm.showLockedItems" ng-change="vm.getItems()"><span translate="Infusion.LockedItems"></span></label>
+        <label><input class="controls" type="checkbox" id="infuse-all" ng-model="vm.getAllItems" ng-change="vm.getItems()"><span ng-i18next="Infusion.ShowItems"></span></label>
+        <label><input class="controls" type="checkbox" id="infuse-locked" ng-model="vm.showLockedItems" ng-change="vm.getItems()"><span ng-i18next="Infusion.LockedItems"></span></label>
       </div>
-    </div>
```
Comment (bhollis) [inline src/scripts/infuse/infuse.html]: Something happened with the HTML in this page - I think it's missing a tag now (or nested differently).

File: src/scripts/move-popup/dimMoveItemProperties.directive.js [hunk 3]
```diff
@@ -154,7 +154,7 @@ function MoveItemPropertiesCtrl($sce, $q, dimStoreService, dimItemService, dimSe
   };
 
   vm.light = null;
-  vm.classType = '';
+  vm.classType = ' ';
```
Comment (bhollis) [inline src/scripts/move-popup/dimMoveItemProperties.directive.js]: What's this change?

File: src/i18n/dim_de.json [hunk 4]
```diff
@@ -1,481 +1,491 @@
...
-         "LearnMore": "Willst du mehr wissen?"
-      }
-   },
-   "TrialsCard": {
-      "FiveWins": "5 Siege Belohnung (Rüstung)",
-      "SevenWins": "7 Siege Belohnung (Waffe)",
-      "Flawless": "Makellos"
-   }
+    "Bucket": {
+        "Armor": "Rüstung",
```
Comment (delphiactual) [inline src/i18n/dim_de.json]: The download service for the other languages will always output with these no matter how I upload them. So English will be normal and the others will have consistent spacing amongst them.

File: src/scripts/infuse/infuse.html (hunk 2 above)
Comment (delphiactual) [inline src/scripts/infuse/infuse.html]: I fixed the nesting, no tags are missing

**Comment by bhollis** [comment]
Sorry I made more merge conflicts. 😅

File: src/scripts/move-popup/dimMoveItemProperties.directive.js (hunk 3 above)
Comment (delphiactual) [inline src/scripts/move-popup/dimMoveItemProperties.directive.js]: i18next didn't like the empty string, but took a space with no problem, it only affects when no classType is assigned to the object.

**Comment by bhollis** [comment]
BTW, I checked this out and there appear to be a few strings missing (like the new "log out" menu item). Might want to do a once-over.

File: src/scripts/compare/compare.html [hunk 5]
```diff
@@ -1,8 +1,8 @@
 <div id="loadout-drawer" ng-if="vm.show">
   <div class="compare-options">
-    <label ng-if="vm.archeTypes.length > 1" class="dim-button" ng-click="vm.compareSimilar('archetype')" translate="{{ vm.compare.location.inWeapons ? 'Compare.Archetype' : 'Compare.Splits' }}" translate-values="{ quantity:  vm.archeTypes.length }"></label>
-    <label ng-if="vm.similarTypes.length > 1" class="dim-button" ng-click="vm.compareSimilar()" translate="Compare.All" translate-values="{ type: vm.compare.typeName, quantity: vm.similarTypes.length }"></label>
-    <label class="dim-button" ng-click="vm.cancel()" translate="Compare.Close"></label>
+    <label ng-if="vm.archeTypes.length > 1" class="dim-button" ng-click="vm.compareSimilar('archetype')" ng-i18next="[i18next]({ quantity: vm.archeTypes.length }){{ vm.compare.location.inWeapons ? 'Compare.Archetype' : 'Compare.Splits' }}"></label>
+    <label ng-if="vm.similarTypes.length > 1" class="dim-button" ng-click="vm.compareSimilar()" ng-i18next="[i18next]({ type: vm.compare.typeName, quantity: vm.similarTypes.length })Compare.All"></label>
```
Comment (delphiactual) [inline src/scripts/compare/compare.html]: which line is bizarre? 'archetypes' or 'similarTypes'? there is probably a cleaner way to write this line but I tried to keep things as much the same as possible...

Comment (kyleshay) [inline src/scripts/compare/compare.html]: he may be referring to the sting format *[i18next]({ ... })({ ... })* that the directive expects

**Pull request merged**